src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs

```diff
@@ -75,6 +74,7 @@ public void SetMemory(OwnedMemory<byte> ownedMemory, int start, int end, bool re
         public void ResetMemory()
         {
             _ownedMemory.Release();
+            _ownedMemory.Dispose();
```

Comment:
This isn't correct. We just need to release.

Reply:
Are you sure? Pipe is the thing that got OM from the pool, so it is its responsibility to dispose it.

Reply:
But even your PR descriptions says "Releasing ArrayMemoryPool block causes it to be disposed and returned to the pool."

Reply:
But calling dispose would also validate that there are no other owners left.

This is beneficial in the case of pipelines. Imagine a situation when someone reads ROS from the pipe, get's memory from it and calls Retain. Now when he tries to advance, Dispose would throw and avoid use after free bugs.

Reply:
On the other hand, we can decide to allow consumers to Retain memory received from pipe and calling dispose here would break this feature.

Reply:
I think we should just release and not dispose